src/pubsub.rs

```diff
@@ -274,45 +270,105 @@ impl PubSub {
-        match bincode::serialize_into(&mut self.serialize_buf, item) {
-            Ok(()) => tx_mapping.enqueue_bulk_bytes(&[(topic.as_ref(), &self.serialize_buf)]),
+        if self.serialize_bufs.is_empty() {
+            self.serialize_bufs.push(Vec::new());
+        }
+
+        self.serialize_bufs[0].clear();
+        match bincode::serialize_into(&mut self.serialize_bufs[0], item) {
+            Ok(()) => tx_mapping.enqueue_bulk_bytes(&[(topic.as_ref(), &self.serialize_bufs[0])]),
             Err(_) => return,
         };
     }
 
-    /// Dequeue a message from the broker, returning the topic and the message
-    /// in a newly-allocated Vec<u8>
-    pub fn dequeue_bytes(&mut self) -> Option<(&str, &[u8])> {
-        let rx_mapping = self.rx_mapping.as_mut()?;
-        
-        self.dequeue_topic.clear();
-        self.dequeue_buf.clear();
-        
-        let mut data = [(Some(&mut self.dequeue_topic), &mut self.dequeue_buf)];
-        let dequeued = rx_mapping.dequeue_bulk_bytes(&mut data, true);
-        
-        if dequeued != 1 {
-            return None;
+    /// Enqueue multiple messages to be sent to the broker under their respective topics.
+    /// 
+    /// Each message is serialized using bincode. The messages are provided as
+    /// tuples of (topic, item).
+    pub fn enqueue_bulk_type<S, T>(&mut self, items: &[(S, &T)])
+    where
+        S: AsRef<str>,
+        T: serde::Serialize,
+    {
+        let tx_mapping = match self.tx_mapping.as_mut() {
+            Some(m) => m,
+            None => return,
+        };
+
+        // resize buffer vec if needed
+        if self.serialize_bufs.len() < items.len() {
+            self.serialize_bufs.resize_with(items.len(), Vec::new);
+        }
+
+        for (i, (_, item)) in items.iter().enumerate() {
+            let buf = &mut self.serialize_bufs[i];
+            buf.clear();
+            if bincode::serialize_into(buf, *item).is_err() {
+                return;
+            }
+        }
+
+        let pairs: Vec<(&str, &[u8])> = items.iter()
+            .zip(self.serialize_bufs.iter())
+            .map(|((topic, _), data)| (topic.as_ref(), data.as_slice()))
+            .collect();
+
+        tx_mapping.enqueue_bulk_bytes(&pairs);
+    }
+
+    /// Dequeue multiple messages from the broker, returning vec of (topic, payload) tuples.
+    /// Returns empty vec if no messages were dequeued.
+    pub fn dequeue_bulk_bytes(&mut self, count: usize) -> &Vec<(String, Vec<u8>)> {
+        let rx_mapping = match self.rx_mapping.as_mut() {
+            Some(m) => m,
+            None => {
+                self.dequeue_bufs.clear();
+                return &self.dequeue_bufs;
+            }
+        };
+    
+        // ensure tmp buffers are big enough
+        if self.dequeue_bufs.len() < count {
+            self.dequeue_bufs.resize_with(count, || (String::new(), Vec::new()));
         }
+        // but not too big
+        self.dequeue_bufs.truncate(count);
```

Comment:
I wouldn't truncate the buffers as a later call might use the temp bufs, maybe passing a view like `&mut items[..count]` and making the signature on `dequeue_bulk_bytes` be `items: &mut [(String, Vec<u8>)]` would work?

Reply:
made these changes so we can avoid resizing frequently